src/devices/Arduino/ArduinoGpioControllerDriver.cs

```diff
@@ -0,0 +1,305 @@
+            _supportedPinConfigurations = supportedPinConfigurations ?? throw new ArgumentNullException(nameof(supportedPinConfigurations));
+            _callbackContainers = new Dictionary<int, CallbackContainer>();
+            _waitForEventResetEvent = new AutoResetEvent(false);
+            _callbackContainersLock = new object();
+            _pinModes = new ConcurrentDictionary<int, PinMode>();
+
+            PinCount = _supportedPinConfigurations.Count;
+            _arduinoBoard.Firmata.DigitalPortValueUpdated += FirmataOnDigitalPortValueUpdated;
+        }
+
+        protected override int PinCount { get; }
+
+        /// <summary>
+        /// Arduino does not distinguish between logical and physical numbers, so this always returns identity
+        /// </summary>
+        protected override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber)
+        {
+            return pinNumber;
+        }
+
+        protected override void OpenPin(int pinNumber)
+        {
+        }
+
+        protected override void ClosePin(int pinNumber)
+        {
+            _pinModes.TryRemove(pinNumber, out _);
+        }
+
+        protected override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            SupportedMode firmataMode;
+            switch (mode)
+            {
+                case PinMode.Output:
+                    firmataMode = SupportedMode.DIGITAL_OUTPUT;
+                    break;
+                case PinMode.InputPullUp:
+                    firmataMode = SupportedMode.INPUT_PULLUP;
+                    break;
+                case PinMode.Input:
+                    firmataMode = SupportedMode.DIGITAL_INPUT;
+                    break;
+                default:
+                    throw new NotSupportedException($"Mode {mode} is not supported for this operation");
+            }
+
+            var pinConfig = _supportedPinConfigurations.FirstOrDefault(x => x.Pin == pinNumber);
+            if (pinConfig == null || !pinConfig.PinModes.Contains(firmataMode))
+            {
+                throw new NotSupportedException($"Mode {mode} is not supported on Pin {pinNumber}.");
+            }
+
+            _arduinoBoard.Firmata.SetPinMode(pinNumber, firmataMode);
+
+            // Cache this value. Since the GpioController calls GetPinMode when trying to write a pin (to verify whether it is set to output),
+            // that would be very expensive here. And setting output pins should be cheap.
+            _pinModes[pinNumber] = mode;
+        }
+
+        protected override PinMode GetPinMode(int pinNumber)
+        {
+            if (_pinModes.TryGetValue(pinNumber, out var existingValue))
+            {
+                return existingValue;
+            }
+
+            SupportedMode mode = _arduinoBoard.Firmata.GetPinMode(pinNumber);
+
+            PinMode ret;
+            switch (mode)
+            {
+                case SupportedMode.DIGITAL_OUTPUT:
```

Comment:
can a pin be digital or analog? if so should you be checking for flags rather than value?

Reply:
This is the current mode of a pin (as opposed to the pin capabilities). This can have one value at a time only.